registry/storage/driver/middleware/alicdn/middleware.go

```diff
@@ -0,0 +1,110 @@
+package middleware
+
+import (
+	"fmt"
+	"hash/fnv"
+
+	"github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	storagemiddleware "github.com/docker/distribution/registry/storage/driver/middleware"
+	"net/url"
+	"time"
+)
+
+type aliCDNStorageMiddleware struct {
+	storagedriver.StorageDriver
+	scheme string
+	host   string
+}
+
+func newAliCDNStorageMiddleware(storageDriver storagedriver.StorageDriver, options map[string]interface{}) (storagedriver.StorageDriver, error) {
+	o, ok := options["baseurl"]
+	if !ok {
+		return nil, fmt.Errorf("no baseurl provided")
+	}
+	b, ok := o.(string)
+	if !ok {
+		return nil, fmt.Errorf("baseurl must be a string")
+	}
+	u, err := url.Parse(b)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse alicdn baseurl: %s", b)
+	}
+	if u.Scheme == "" {
+		return nil, fmt.Errorf("no scheme specified for alicdn baseurl")
+	}
+	if u.Host == "" {
+		return nil, fmt.Errorf("no host specified for alicdn baseurl")
+	}
+	return &aliCDNStorageMiddleware{
+		StorageDriver: storageDriver,
+		scheme:        u.Scheme,
+		host:          u.Host,
+	}, nil
+}
+
+func (cdn *aliCDNStorageMiddleware) h(data string) int {
+	h := fnv.New32a()
+	h.Write([]byte(data))
+	return int(h.Sum32())
+}
+
+func (cdn *aliCDNStorageMiddleware) expireOf(now time.Time, path string) time.Time {
+
+	// Every file has it's own expire point based on hash of the path,
```

Comment:
Is this algorithm documented somewhere or is this custom logic here to calculating?